pkg/api/scylla/v1alpha1/types_scylladbdatacenter.go

```diff
@@ -0,0 +1,528 @@
+	// +optional
+	RackTemplate *RackTemplate `json:"rackTemplate,omitempty"`
+
+	// racks specify the racks in the datacenter.
+	Racks []RackSpec `json:"racks"`
+
+	// disableAutomaticOrphanedNodeReplacement controls if automatic orphan node replacement should be disabled.
+	DisableAutomaticOrphanedNodeReplacement bool `json:"disableAutomaticOrphanedNodeReplacement,omitempty"`
+
+	// minTerminationGracePeriodSeconds specifies minimum duration in seconds to wait before every drained node is
+	// terminated. This gives time to potential load balancer in front of a node to notice that node is not ready anymore
+	// and stop forwarding new requests.
+	// This applies only when node is terminated gracefully.
+	// If not provided, Operator will determine this value.
+	// EXPERIMENTAL. Do not rely on any particular behaviour controlled by this field.
+	// +optional
+	MinTerminationGracePeriodSeconds *int32 `json:"minTerminationGracePeriodSeconds,omitempty"`
+
+	// minReadySeconds specifies the minimum number of seconds for which a newly created ScyllaDB node should be ready
+	// for it to be considered available.
+	// When used to control load balanced traffic, this can give the load balancer in front of a node enough time to
+	// notice that the node is ready and start forwarding traffic in time. Because it all depends on timing, the order
+	// is not guaranteed and, if possible, you should use readinessGates instead.
+	// If not provided, Operator will determine this value.
+	// +optional
+	MinReadySeconds *int32 `json:"minReadySeconds,omitempty"`
+
+	// readinessGates specifies custom readiness gates that will be evaluated for every ScyllaDB Pod readiness.
+	// It's projected into every ScyllaDB Pod as its readinessGate. Refer to upstream documentation to learn more
+	// about readiness gates.
+	// +optional
+	ReadinessGates []corev1.PodReadinessGate `json:"readinessGates,omitempty"`
+}
+
+type ObjectTemplateMetadata struct {
+	// labels specify a custom key value map that gets merged with managed object labels.
+	// +optional
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// annotations specify a custom key value map that gets merged with managed object annotations.
+	// +optional
+	Annotations map[string]string `json:"annotations,omitempty"`
+}
+
+type RackTemplate struct {
+	// nodes specify the desired number of nodes in rack.
+	// +optional
+	Nodes *int32 `json:"nodes"`
+
+	// placement describes restrictions for the nodes ScyllaDB is scheduled on.
+	// +optional
+	Placement *Placement `json:"placement,omitempty"`
+
+	// topologyLabelSelector specifies a label selector which will be used to target nodes at specified topology constraints.
+	// Datacenter topologyLabelSelector is merged with rack topologyLabelSelector and then converted into nodeAffinity
+	// targeting nodes having specified topology.
+	// +optional
+	TopologyLabelSelector map[string]string `json:"topologyLabelSelector,omitempty"`
```

Comment:
[In the proposal](https://github.com/scylladb/scylla-operator/blame/2870781/enhancements/proposals/1978-scylladbdatacenter-api/README.md#L288) and say in DaemonSets this is `*metav1.LabelSelector` - why the change?

Reply:
`metav1.LabelSelector` is a complex field, I had `map[string]string` in mind. Examples presented in proposals are using correct type.

Reply:
don't we want to support set based requirements?
https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements

Reply:
Whole point of topologyLabelSelector was simplicity, as most usecases for topology is setting a single label requirement. Users wanting to use expressions and other complex rules have to use what's available in placement.